update to python ~3.8

adrenen opened the pull request:
Hi frem,
I thought it would be a fun weekend project to update 4ChanBBS to work with python3.  I have enjoyed playing with it off and on and wanted to help drag it into the future lol

To do this, several overall changes were made:
    Use of telnetlib3 and conversion to async,
    Use of html.parser,
    Use of sessions when making requests to the 4chan api (this greatly smooths the browsing experience)

Everything passes rigorous *works on my system* testing.  So please let me know if anything doesn't work quite right.

I might do some further work to convert all the api requests over to async and make the image size dynamically adjust to the client terminal dimensions.   Will have to see how lazy I am feeling.

Commit:
test

Commit:
updated requirements to match updates

Commit:
moved fetching to chanjson and added configureable image sizes

Commit:
updated to use sessions and preloading of boards

Commit:
updated to add new command and aditional settings

Commit:
updated to use html.parser

Commit:
updated to reflect version changes

Commit:
updated to use telnetlib3 and asyncio

Commit:
updated to current board list

carter-yagemann commented:
Hi @adrenen, thanks for taking the time to revamp the code!

A few pieces of feedback before this PR is merged:

1. On Debian, I had to `sudo apt install libaa1` to get the libaa1 Python package to work. The setup steps in the README should be updated to reflect this.
2. Are all the packages in `requirements.txt` necessary? I noticed some new packages like `charset-normalizer` that never seem to get imported. Unless I'm missing something, `requirements.txt` needs to be cleaned up.
3. When I tested the code, characters typed into the prompt weren't echoed back, so I couldn't see the command I was typing. I can't remember if this is the original behavior, but in either case, it should be fixed if possible. Similarly, backspace doesn't work when entering commands.
4. There's room for some simple quality of life improvements: a help/usage command and a message when the user enters an invalid command. These weren't in the old version, but if you want to add them in, it'd be greatly appreciated.

Thanks again!

Commit:
cleaned up requirements

Commit:
updated readme to reflect addition of libaa1

Commit:
added negotiation for echo and backspace.  and started callback for NAWS

Commit:
added support for dynamic image resizing

Commit:
added support for NAWS based dynamic image scaling

Commit:
added settings for NAWS

Commit:
added help messages and error handling for bad commands

Commit:
bug fix for invalid board selection in listthreads

Commit:
updated config for naws and reset debug flag

adrenen commented:
Thanks for the feedback!

1. I have updated the README to include instructions for installing `libaa1`
2. All the packages were not necessary.  tbh, I just did pip freeze > requirements.txt  Got that cleaned up.
3. Updated to ask the client to handle the echo and backspaces and stuff.  This should work for the vast majority of modern clients I think.
4. Added some help messages and error messages for the user.  I dont think my formatting looks great, but they get the job done I guess.

I also added support for dynamic image sizing relative to the client window.  That took some playing with NAWS but it wasn't too bad.

Please let me know if there is anything else I can do!

Commit:
spelling fix

carter-yagemann commented:
Looks good to me, thank you!
